### PR TITLE
Jean-Philippe Gouin's OpenLDAP Helm chart adapted to use the Symas-supplied OpenLDAP container.

### PR DESCRIPTION
**Pull request by gburd**
### Description of the change
[Jean-Philippe Gouin's](https://github.com/jp-gouin/helm-openldap) OpenLDAP [Helm chart](https://helm.sh/) is a good starting point for ours for several reasons, including that he's a VMware employee likely part of the Bitnami acquisition and perhaps one of the people working on their supported OpenLDAP container that we've also forked.

This chart is a balancing act, but ultimately, the goal is to provide our customers and everyone else with something reliable, secure, and maintainable.  It should provide the basis for deploying OpenLDAP into a [Kubernetes](https://kubernetes.io/) cluster to provide a performant, robust (read: replicated, fault-tolerant), scalable cluster.

As we have been supporting OpenLDAP in production for quite a while now (>15y), I suspect some settings, defaults, etc. in this configuration are sub-optimal. Things may have been overlooked; items may need to be updated or removed. This review aims to identify and fix those and any other issues so that our community of users can have confidence in our deliverable.

### CODE REVIEW

File: .bin/local-test.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+#set -x
+
+CERT_DIR=${CERT_DIR:-$(mktemp -d)}
+NAMESPACE=ds
+KIND_CLUSTER_NAME=kind
+
+if ! $(kind get clusters -q | grep $KIND_CLUSTER_NAME > /dev/null 2>&1); then
```
Comment (mistotebe) [inline .bin/local-test.sh:19]: `grep -q`?

File: CODE_OF_CONDUCT.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# Contributor Covenant Code of Conduct
```
Comment (mistotebe) [inline CODE_OF_CONDUCT.md:1]: Did you want to harmonise this with the others?

File: charts/common-1.10.0.tgz
Comment (mistotebe) [inline charts/common-1.10.0.tgz:1]: Is this file generated? If so why is it also stored?

File: templates/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+{{- $configPassword :=  ternary .Values.global.configPassword "%%CONFIG_PASSWORD%%" (empty .Values.global.existingSecret) }}
+{{- $retry := .Values.replication.retry }}
+{{- $timeout := .Values.replication.timeout }}
+{{- $starttls := .Values.replication.starttls }}
+{{- $tls_reqcert := .Values.replication.tls_reqcert }}
+{{- $nodeCount := .Values.replicaCount | int }}
+{{- $port := .Values.global.ldapPort | int }}
+  {{- range $index0 := until $nodeCount }}
+    {{- $index1 := $index0 | add1 }}
+    olcSyncRepl: rid=00{{ $index1 }} provider=ldap://{{ $name }}-{{ $index0 }}.{{ $name }}-headless.{{ $namespace }}.svc.{{ $cluster }}:{{ $port }} binddn="cn={{ $bindDNUser }},cn=config" bindmethod=simple credentials={{ $configPassword }} searchbase="cn=config" type=refreshAndPersist retry="{{ $retry }} +" timeout={{ $timeout }} starttls={{ $starttls }} tls_reqcert={{ $tls_reqcert }}
```
Comment (mistotebe) [inline templates/_helpers.tpl:84]: binddn looks impossible to provide, might also want to decide whether you want to wrap as in the olcSyncrepl below or keep all as one line and use the same style in both places.

File: templates/_helpers.tpl [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+
+{{/*
+Return the list of builtin schema files to mount
+Cannot return list => return string comma separated
+*/}}
+{{- define "openldap.builtinSchemaFiles" -}}
+  {{- $schemas := "" -}}
+  {{- if .Values.replication.enabled -}}
+    {{- $schemas = "syncprov,serverid,csyncprov,rep,bsyncprov,brep,acls" -}}
```
Comment (mistotebe) [inline templates/_helpers.tpl:167]: Do you need to pretend all of these (config changes) are schemas to smuggle them in? If so, you should make changes to the container so there's a proper way of doing this kind of thing and use it here.

File: templates/configmap-env.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  LDAP_GROUP: {{ .Values.group }}
+  {{- end }}
+  LDAP_TLS_CERT_FILE: /opt/bitnami/openldap/certs/tls.crt
+  LDAP_TLS_KEY_FILE: /opt/bitnami/openldap/certs/tls.key
+  LDAP_TLS_CA_FILE: /opt/bitnami/openldap/certs/ca.crt
+  {{- if .Values.global.configUserEnabled }}
+  LDAP_CONFIG_ADMIN_ENABLED: "yes"
+  LDAP_CONFIG_ADMIN_USERNAME: "{{ .Values.global.configUser }}"
+  {{- else }}
+  LDAP_CONFIG_ADMIN_ENABLED: "no"
```
Comment (mistotebe) [inline templates/configmap-env.yaml:38]: Have I missed something or is cn=config replication set up even if this is set to `no`? Is there a reason for that?

File: .bin/local-test.sh (hunk 1 above)
Comment (gburd) [reply .bin/local-test.sh:19]: Fixed, despite that only being in a test script used locally only. :)

File: templates/_helpers.tpl (hunk 4 above)
Comment (gburd) [reply templates/_helpers.tpl:167]: There are three ways before `slapd` starts and listens on TCP port(s) making it "available for service" in the eyes of Kubernetes and one after:
* extra config
* custom schema
* custom schemas
and after
* ldap custom init scripts

What's better here?  I'd imagine that having these loaded into the server before it opens the TCP port, so traffic doesn't start arriving when replication isn't setup but you tell me.

Comment (mistotebe) [reply templates/_helpers.tpl:167]: Depends, do you want to rely on the API that the container provides? Then extend that in ways that make sense (at least) for your usecase. Also if the server appears live in Kubernetes before "custom init scripts" are even run, then that sounds like a problem anyway.

File: templates/_helpers.tpl (hunk 3 above)
Comment (gburd) [reply templates/_helpers.tpl:84]: fixed

File: templates/_helpers.tpl (hunk 4 above)
Comment (gburd) [reply templates/_helpers.tpl:167]: fixed, they happen much later in a more obvious location now

**Pull request closed without merging**